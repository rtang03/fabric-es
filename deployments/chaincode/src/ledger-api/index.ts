export * from './statelist';
export * from './utils';
export * from './commit';
export * from './privateStateList';
