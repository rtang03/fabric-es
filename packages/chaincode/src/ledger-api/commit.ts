import { makeKey } from './index';

export interface BaseEvent {
  readonly type: string;
  payload: any;
}

export class Commit<T extends BaseEvent = any> {
  key: string;
  id: string;
  entityName: string;
  version: number;
  commitId: string;
  committedAt: string;
  entityId: string;
  events: BaseEvent[];

  constructor({
    id,
    entityName,
    commitId,
    version,
    committedAt,
    entityId,
    events
  }: {
    id: string;
    entityName: string;
    commitId: string;
    version: number;
    committedAt: string;
    entityId: string;
    events: any;
  }) {
    this.key = makeKey([entityName, id, commitId]);
    this.id = id;
    this.entityName = entityName;
    this.version = version;
    this.commitId = commitId;
    this.committedAt = committedAt;
    this.entityId = entityId;
    this.events = events;
  }
}
