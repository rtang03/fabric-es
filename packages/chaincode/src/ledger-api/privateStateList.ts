import { Context } from 'fabric-contract-api';
import { Iterators } from 'fabric-shim';
import { omit } from 'lodash';
import { serialize, splitKey } from '.';
import { Commit } from '..';

export class PrivateStateList {
  constructor(public ctx: Context, public name: string) {}

  async getQueryResult(collection: string, attributes: string[]) {
    // todo: this is workaround to known bug FAB-14216. This will not work after the bug is fixexd in V2.
    const iterator: Iterators.StateQueryIterator = await this.ctx.stub
      .getPrivateDataByPartialCompositeKey(collection, 'entities', attributes)
      .then((data: any) => data.iterator);
    const result: any = {};
    while (true) {
      const { value, done } = await iterator.next();
      if (value && value.value.toString()) {
        const commit = JSON.parse(value.value.toString('utf8'));
        result[commit.commitId] = omit(commit, 'key');
      }
      if (done) {
        await iterator.close();
        return Buffer.from(JSON.stringify(result));
      }
    }
  }

  async addState(collection: string, commit: Commit) {
    await this.ctx.stub.putPrivateData(
      collection,
      this.ctx.stub.createCompositeKey(this.name, splitKey(commit.key)),
      serialize(commit)
    );
  }

  async getState(collection: string, key: string): Promise<Commit> {
    const data = await this.ctx.stub.getPrivateData(
      collection,
      this.ctx.stub.createCompositeKey(this.name, splitKey(key))
    );
    if (data.toString()) {
      return JSON.parse(data.toString());
    } else return Object.assign({});
  }

  async deleteState(collection: string, { key }: Commit) {
    await this.ctx.stub.deletePrivateData(collection, this.ctx.stub.createCompositeKey(this.name, splitKey(key)));
  }
}
