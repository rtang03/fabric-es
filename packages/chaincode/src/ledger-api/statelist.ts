import util from 'util';
import { Context } from 'fabric-contract-api';
import { keys, omit } from 'lodash';
import { Commit } from '..';
import { serialize, splitKey } from '.';

export class StateList {
  constructor(public ctx: Context, public name: string) {}

  async getQueryResult(attributes: string[], plainObject?: boolean) {
    const promises = this.ctx.stub.getStateByPartialCompositeKey('entities', attributes);
    const result: any = {};

    try {
      for await (const res of promises) {
        const commit = JSON.parse(res.value.toString());
        result[commit.commitId] = omit(commit, 'key');
      }
    } catch (e) {
      console.error(e);
      throw new Error(util.format('fail to getQueryResult, %j', e));
    }
    return plainObject ? result : Buffer.from(JSON.stringify(result));
  }

  async addState(commit: Commit) {
    await this.ctx.stub.putState(this.ctx.stub.createCompositeKey(this.name, splitKey(commit.key)), serialize(commit));
  }

  async getState(key): Promise<Commit> {
    let result;

    const data = await this.ctx.stub.getState(this.ctx.stub.createCompositeKey(this.name, splitKey(key)));

    try {
      result = data.toString() ? JSON.parse(data.toString()) : Object.assign({});
    } catch (e) {
      console.error(e);
      throw new Error(util.format('fail to parse data, %j', e));
    }

    return result;
  }

  async deleteState(commit: Commit) {
    await this.ctx.stub.deleteState(this.ctx.stub.createCompositeKey(this.name, splitKey(commit.key)));
  }

  async deleteStateByEnityId(attributes: string[]) {
    const promises = this.ctx.stub.getStateByPartialCompositeKey('entities', attributes);
    const result = {};
    try {
      for await (const res of promises) {
        const { key, commitId } = JSON.parse(res.value.toString());
        await this.ctx.stub.deleteState(this.ctx.stub.createCompositeKey('entities', splitKey(key)));
        result[commitId] = {};
      }
    } catch (e) {
      console.error(e);
      throw new Error(util.format('fail to deleteStateByEnityId, %j', e));
    }

    return Buffer.from(
      JSON.stringify({
        status: 'SUCCESS',
        message: `${keys(result).length} record(s) deleted`,
        result
      })
    );
  }

}
