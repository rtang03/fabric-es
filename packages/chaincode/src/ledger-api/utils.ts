import { assign } from 'lodash';
import { Commit } from '..';

export const splitKey = (key: string) => key.split('~');

export const makeKey = (keyParts: any[]) =>
  keyParts.map(part => JSON.stringify(part)).join('~');

export const serialize = object => Buffer.from(JSON.stringify(object));

export const toRecord = (commit: Partial<Commit>) =>
  assign({}, { [commit.commitId]: commit });

export const createInstance = ({
  id,
  entityName,
  version,
  events
}: {
  id: string;
  entityName: string;
  version: string;
  events: any[];
}) => {
  const now = Date.now();
  const date = new Date(now).toISOString().replace(/[^0-9]/g, '');
  const commitId = `${date}`;
  const committedAt = now.toString();

  return new Commit({
    id,
    entityName,
    commitId,
    committedAt,
    version: parseInt(version, 10),
    events,
    entityId: id
  });
};
