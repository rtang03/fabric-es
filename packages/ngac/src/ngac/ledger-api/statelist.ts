import { Context } from 'fabric-contract-api';

export const splitKey = (key: string) => key.split(':');

export const makeKey = (keyParts: any[]) =>
  keyParts.map(part => JSON.stringify(part)).join(':');

export const serialize = object => Buffer.from(JSON.stringify(object));

export const getStateList: (name: string, context: Context) => any = (
  name,
  { stub }
) => ({
  getQueryResult: async (attributes: string[]) => {
    const iterator = await stub.getStateByPartialCompositeKey(name, attributes);
    const result = {};
    while (true) {
      const { value, done } = await iterator.next();
      if (value && value.value.toString()) {
        const json = JSON.parse((value.value as Buffer).toString('utf8'));
        result[json.key] = json;
      }
      if (done) {
        await iterator.close();
        return Buffer.from(JSON.stringify(result));
      }
    }
  },
  addState: async json =>
    await stub.putState(
      stub.createCompositeKey(name, splitKey(json.key)),
      serialize(json)
    ),
  getState: async key => {
    const data = await stub.getState(
      stub.createCompositeKey(name, splitKey(key))
    );
    return data.toString() ? JSON.parse(data.toString()) : {};
  },
  deleteStateByKey: async key =>
    await stub.deleteState(stub.createCompositeKey(name, splitKey(key)))
});
