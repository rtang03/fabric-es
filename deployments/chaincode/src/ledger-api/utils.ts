import { assign } from 'lodash';
import { Commit } from './commit';

export const splitKey = (key: string) => key.split('~');

export const makeKey = (keyParts: any[]) => keyParts.map(part => JSON.stringify(part)).join('~');

export const serialize = object => Buffer.from(JSON.stringify(object));

export const toRecord = (commit: Partial<Commit>) => assign({}, { [commit.commitId]: commit });

export const createCommitId = () => `${new Date(Date.now()).toISOString().replace(/[^0-9]/g, '')}`;

export const createInstance = (option: {
  id: string;
  entityName: string;
  version: string;
  mspId: string;
  events: any[];
  commitId: string;
}) =>
  new Commit({
    id: option.id,
    entityName: option.entityName,
    commitId: option.commitId,
    version: parseInt(option.version, 10),
    mspId: option.mspId,
    events: option.events,
    entityId: option.id
  });

// type guard for transient data
export const isEventArray = (value: unknown): value is { type: string; lifeCycle?: number; payload?: any }[] =>
  Array.isArray(value) && value.every(item => typeof item.type === 'string');
