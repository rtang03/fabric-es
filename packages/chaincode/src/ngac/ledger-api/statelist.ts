import { Context } from 'fabric-contract-api';
import { Attribute, Policy, Resource } from '../types';
import { splitKey } from '../utils';

const serialize = object => Buffer.from(JSON.stringify(object));

const stateList: <T extends Attribute | Attribute[] | Resource | Policy = any>(
  namespace: string,
  context: Context
) => {
  getQueryResult: (keyparts: string[]) => Promise<any[]>;
  addState: (key: string, item: T) => Promise<T>;
  getState: (key: string) => Promise<T>;
  deleteStateByKey: (key: string) => Promise<string>;
  deleteStatesByKeyRange: (keyparts: string[]) => Promise<string[]>;
} = <T extends Attribute | Attribute[] | Resource | Policy = any>(namespace, { stub }) => ({
  getQueryResult: async keyparts => {
    const iterator = await stub.getStateByPartialCompositeKey(namespace, keyparts);
    const result: T[] = [];
    while (true) {
      const { value, done } = await iterator.next();
      if (value && value.value.toString()) {
        const json: T = JSON.parse(value.value.toString('utf8'));
        result.push(json);
      }
      if (done) {
        await iterator.close();
        return result;
      }
    }
  },
  addState: async (key, item) => {
    await stub.putState(stub.createCompositeKey(namespace, splitKey(key)), serialize(item));
    return item;
  },
  getState: async key => {
    const data = await stub.getState(stub.createCompositeKey(namespace, splitKey(key)));
    return data.toString() ? JSON.parse(data.toString()) : {};
  },
  deleteStateByKey: async inputkey => {
    const key = stub.createCompositeKey(namespace, splitKey(inputkey));
    await stub.deleteState(key);
    return key;
  },
  deleteStatesByKeyRange: async keyparts => {
    const iterator = await stub.getStateByPartialCompositeKey(namespace, keyparts);
    const result = [];
    while (true) {
      const { value, done } = await iterator.next();
      if (value && value.value.toString()) {
        const item = JSON.parse(value.value.toString('utf8'));
        await stub.deleteState(stub.createCompositeKey(namespace, splitKey(item.key)));
        result.push(item.key);
      } else return [];

      if (done) {
        await iterator.close();
        return result;
      }
    }
  }
});

export default stateList;
