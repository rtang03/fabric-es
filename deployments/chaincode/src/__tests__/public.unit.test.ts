import { values } from 'lodash';
import { Commit, EventStore } from '..';
import { StateList } from '../ledger-api';

const ctx: any = {
  stub: {
    createCompositeKey: jest.fn(),
    deleteState: jest.fn(),
    getState: jest.fn(),
    putState: jest.fn(),
    setEvent: jest.fn(),
    getStateByPartialCompositeKey: jest.fn(),
    getCreator: jest.fn()
  },
  clientIdentity: { getID: jest.fn() }
};
const context = {
  stateList: new StateList(ctx, 'entities'),
  ...ctx
};

ctx.stub.createCompositeKey.mockResolvedValue('entities"en""entId""2019"');
ctx.stub.putState.mockResolvedValue(Buffer.from(''));
ctx.stub.setEvent.mockImplementation((name, args) => console.log(`Event sent: ${name}: ${args}`));
ctx.stub.getCreator.mockImplementation(() => { return { 'mspid': 'Org1MSP' }; });
ctx.clientIdentity.getID.mockImplementation(() => 'Org1MSP');

const cc = new EventStore(context);
const entityName = 'cc_test';
const id = 'cc_01';
const entityId = id;
const version = '0';
const events = [{ type: 'mon', payload: { name: 'jun' } }];
const eventStr = JSON.stringify(events);
const commitId = '123';
const committedAt = '2019';
const value = JSON.stringify({
  key: '123', // any non-null string
  commitId,
  committedAt,
  version,
  entityName,
  id,
  entityId,
  events
});

ctx.stub.getStateByPartialCompositeKey.mockImplementation(() => {
  let counter = 1;
  return {
    [Symbol.asyncIterator]: () => ({
      next: () => {
        if (counter > 0) {
          counter--;
          return Promise.resolve({ value: { value }, done: false });
        } else return Promise.resolve({ done: true });
      }
    })
  };
});
ctx.stub.getState.mockResolvedValue(value);

describe('Chaincode Tests', () => {
  it('should instantiate', async () => cc.Init(context).then(response => expect(response).toEqual('Init Done')));

  it('should createCommit', async () =>
    cc
      .createCommit(context, entityName, id, version, eventStr, commitId)
      .then<Commit>((response: any) => values(JSON.parse(response))[0])
      .then(({ id, entityName, version, entityId }) => ({
        id,
        entityName,
        version,
        entityId,
        events
      }))
      .then(commit => expect(commit).toMatchSnapshot()));

  it('should queryByEntityName', async () =>
    cc
      .queryByEntityName(context, entityName)
      .then(response => JSON.parse(response))
      .then(response => expect(response).toMatchSnapshot()));


  it('should queryByEntityId', async () =>
    cc
      .queryByEntityId(context, entityName, id)
      .then<Record<string, Commit>>((response: any) => JSON.parse(response))
      .then(response => expect(response).toMatchSnapshot()));

  it('should queryByEntityIdCommitId', async () =>
    cc
      .queryByEntityIdCommitId(context, entityName, id, commitId)
      .then<Record<string, Commit>>((response: any) => JSON.parse(response))
      .then(response => expect(response).toMatchSnapshot()));

  it('should deleteByEntityIdCommitId', async () =>
    cc
      .deleteByEntityIdCommitId(context, entityName, id, commitId)
      .then<Record<string, Commit>>((response: any) => JSON.parse(response))
      .then(({ status }) => expect(status).toBe('SUCCESS')));

  it('should deleteByEntityId', async () =>
    cc
      .deleteByEntityId(context, entityName, id)
      .then<Record<string, Commit>>((response: any) => JSON.parse(response))
      .then(({ status }) => expect(status).toBe('SUCCESS')));

});
