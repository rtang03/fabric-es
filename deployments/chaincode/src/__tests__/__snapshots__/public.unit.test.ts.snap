// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chaincode Tests should createCommit 1`] = `
Object {
  "entityId": "cc_01",
  "entityName": "cc_test",
  "events": Array [
    Object {
      "payload": Object {
        "name": "jun",
      },
      "type": "mon",
    },
  ],
  "id": "cc_01",
  "version": 0,
}
`;

exports[`Chaincode Tests should queryByEntityId 1`] = `
Object {
  "123": Object {
    "commitId": "123",
    "committedAt": "2019",
    "entityId": "cc_01",
    "entityName": "cc_test",
    "events": Array [
      Object {
        "payload": Object {
          "name": "jun",
        },
        "type": "mon",
      },
    ],
    "id": "cc_01",
    "version": "0",
  },
}
`;

exports[`Chaincode Tests should queryByEntityIdCommitId 1`] = `
Object {
  "123": Object {
    "commitId": "123",
    "committedAt": "2019",
    "entityId": "cc_01",
    "entityName": "cc_test",
    "events": Array [
      Object {
        "payload": Object {
          "name": "jun",
        },
        "type": "mon",
      },
    ],
    "id": "cc_01",
    "version": "0",
  },
}
`;

exports[`Chaincode Tests should queryByEntityName 1`] = `
Object {
  "123": Object {
    "commitId": "123",
    "committedAt": "2019",
    "entityId": "cc_01",
    "entityName": "cc_test",
    "events": Array [
      Object {
        "payload": Object {
          "name": "jun",
        },
        "type": "mon",
      },
    ],
    "id": "cc_01",
    "version": "0",
  },
}
`;
